[crypto] allow key reference usage for ECDSA

hemanth-silabs opened the pull request:
Add new class to handle ECDSA using key refs.
Update SRP to use key refs.

hemanth-silabs commented:
@jwhui @abtink @Damian-Nordic Can you please review this PR? Based on the previous discussion, I have left the `KeyPair` as it is and added a new class `KerPairAsRef` for use with PSA.

hemanth-silabs commented:
@jwhui Can you please add reviewers?

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +217,102 @@ class P256
+    /**
+     * This class represents a key pair (public and private keys) as a PSA KeyRef.
+     *
+     * The key pair is stored using Distinguished Encoding Rules (DER) format (per RFC 5915).
```
Remove this line? Not relevant/correct anymore?

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on include/openthread/platform/crypto.h:
```diff
@@ -174,6 +177,19 @@ typedef struct otPlatCryptoEcdsaKeyPair
+ */
+typedef struct otPlatCryptoEcdsaKeyRef
+{
+    uint32_t mKeyRef;
```
Use `otCryptoKeyRef` instead of `uint32_t`?

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on include/openthread/platform/crypto.h:
```diff
@@ -174,6 +177,19 @@ typedef struct otPlatCryptoEcdsaKeyPair
+typedef struct otPlatCryptoEcdsaKeyRef
+{
+    uint32_t mKeyRef;
+} otPlatCryptoEcdsaKeyRef;
```
Do we need `struct`?  We can use:
```
typedef otCryptoKeyRef  otPlatCryptoEcdsaKeyRef
```

We already do the same for Network Key:
```
typedef otCryptoKeyRef otNetworkKeyRef; ///< Reference to Key
```

hemanth-silabs replied:
Yes, we can do that. I wanted to keep the keypairref as close to keypair as possible, and that is why I made it a struct.

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -164,6 +171,12 @@ class P256
+         * This method clears the keypair stored.
+         *
+         */
+        void Clear(void) { memset(mDerBytes, 0, kMaxDerSize); }
```
If we use `Settings::Delete<Key>`  then we won't need this

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +217,102 @@ class P256
+         * @retval kErrorFailed       Failed to import the key.
+         *
+         */
+        Error ImportKeyPair(KeyPair aKeyPair)
```
```suggestion
        Error ImportKeyPair(const KeyPair &aKeyPair)
```

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +217,102 @@ class P256
+
+        /**
+         * This method imports a new keypair into PSA ITS.
+         *
```
Add doc for `aKeyPair` input?

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +217,102 @@ class P256
+    public:
+        /**
+         * This constructor initializes a `KeyPairAsRef`.
+         *
```
Document `aKeyRef` input?

hemanth-silabs replied:
Will push a fix in next commit.

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +217,102 @@ class P256
+
+        /**
+         * This method sets the Key reference.
+         *
```
Doc the input `aKeyRef`?

abtink commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -242,6 +354,27 @@ class P256
+         * @retval kErrorNoBufs        Failed to allocate buffer for signature verification
+         *
+         */
+        Error VerifyUsingKeyRef(uint32_t aKeyRef, const Sha256::Hash &aHash, const Signature &aSignature) const
```
Where do we use this method?
Do we need this? 
This is in `PublicKey` class which I guess don't need to be saved in PSA?

hemanth-silabs replied:
Yes, we do need it. ECDSA can be used by the user for app usecases and it is useful to have verify method along with Sign.

hemanth-silabs replied:
Ok, I can add a public facing API to use this API. But I do see a use for this API.

abtink replied:
The only place we use `Ecdsa::Verify()` today is in `Srp::Server` where we verify the signature of a received "SRP Update" message. In this case, the public key is itself read from the message and won't be saved in the PSA as a key ref (`Srp::Server` only knows the public key which is not security critical).

hemanth-silabs replied:
Ok, in this case, can we have a abstraction for verify, but remove it from the core header? This way anyone can use the API to verify, but we dont need to maintain it in the core.

abtink reviewed (commented):
Thanks @hemanth-silabs.
Some questions/suggestions below.

hemanth-silabs commented on src/core/net/srp_client.cpp:
```diff
@@ -885,6 +890,34 @@ Error Client::PrepareUpdateMessage(Message &aMessage)
+    Error                        error = kErrorNone;
+    Crypto::Ecdsa::P256::KeyPair aKeyPair;
+
+    SuccessOrExit(Crypto::Storage::HasKey(aKeyRef.GetKeyRef()) == true);
```
If the key does not exist, we check for literal key and if that does not exist as well, we generate and import.

hemanth-silabs commented on src/core/net/srp_client.cpp:
```diff
@@ -885,6 +890,34 @@ Error Client::PrepareUpdateMessage(Message &aMessage)
+            aKeyPair.Clear();
+            IgnoreError(Get<Settings>().Save<Settings::SrpEcdsaKey>(aKeyPair));
+            ExitNow();
+        }
```
Good Point.

hemanth-silabs replied:
Will push a fix in next commit.

hemanth-silabs commented on src/core/net/srp_client.cpp:
```diff
@@ -885,6 +890,34 @@ Error Client::PrepareUpdateMessage(Message &aMessage)
+        {
+            SuccessOrExit(error = aKeyRef.ImportKeyPair(aKeyPair));
+            aKeyPair.Clear();
+            IgnoreError(Get<Settings>().Save<Settings::SrpEcdsaKey>(aKeyPair));
```
Sorry, missed that.

hemanth-silabs replied:
Will push a fix in next commit.

hemanth-silabs commented:
@abtink Thank you for the review. I have pushed the changes and addressed some of your questions. Please have a look.

hemanth-silabs commented:
@Damian-Nordic Can you have a look at this PR? Sorry for re-requesting again, had to change the design based on reviews.

Damian-Nordic commented on src/core/net/srp_client.cpp:
```diff
@@ -885,6 +890,34 @@ Error Client::PrepareUpdateMessage(Message &aMessage)
+    Error                        error = kErrorNone;
+    Crypto::Ecdsa::P256::KeyPair keyPair;
+
+    VerifyOrExit(!Crypto::Storage::HasKey(aKeyRef.GetKeyRef()));
```
The problem with this is that OT factory reset will not remove this key, unless `otPlatSettingsWipe` does it, right? We either need `otPlatSettingsWipe` documented that it should also wipe PSA-stored keys or have an alternative crypto API for wiping the keys. Any thoughts?

abtink replied:
I fully agree. I think we save other keys and info in PSA for Thread network which should be wiped as well on a factory reset. 

I would suggest adding a new crypto platform API for this and then we can call it from `FactoryReset()` method.

hemanth-silabs replied:
Funny you should say that, I am parallely working on an issue where otInstanceFinalize does not clear the PSA keys at all. It is fine in non-PSA use cases, where static buffers are overwritten, but in PSA we will need to clear these manually. I have experimental changes in my local branch, where I have added a new API in key manager and sub mac to clear the keys.

what are your thoughts on this? If it is OK, I can bring those changes to this PR as well.

hemanth-silabs replied:
I dont think this can be done in a platform API, because we will need to clear the volatile keys as well, And platform does not store the key refs for these. So either we need to add a new interface in key_manager to clear the keys, or update documentation for `otInstanceFinalize` to istruct this should be followed by a power cycle.

Damian-Nordic replied:
I just realized that for SRP not clearing the key on the factory reset is a severe issue because we will not create a new identity. I didn't thought about volatile keys though so if that requires more changes, a separate PR makes sense to me as well. Whatever works for you better :)

Damian-Nordic commented on src/core/net/srp_client.cpp:
```diff
@@ -1418,7 +1456,11 @@ Error Client::AppendSignature(Message &aMessage, Info &aInfo)
 
     sha256.Finish(hash);
+#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
+    SuccessOrExit(error = aInfo.mKeyRef.Sign(hash, signature));
```
nit: Some of these `#if`s could be avoided if `mKeyRef` was renamed to `mKeyPair` (which is OK since both are mutually exclusive).

hemanth-silabs replied:
That is true, but I opted this way deliberately to let the user know that we are dealing with keyrefs here.

Damian-Nordic replied:
oh, I see.

Damian-Nordic reviewed (commented):
@hemanth-silabs No problem, I found some potential issues that you may take into account.

abtink reviewed (commented):
Sorry! Just noticed! I forgot to post these two comment :)

hemanth-silabs commented on include/openthread/platform/crypto.h:
```diff
@@ -174,6 +177,12 @@ typedef struct otPlatCryptoEcdsaKeyPair
+ * This datatype represents the keypair as PSA key reference.
+ *
+ */
+typedef otCryptoKeyRef otPlatCryptoEcdsaKeyRef;
```
I am following the NetworkKeyRef route here. But yes, it does make sense to just reuse otCryptoKeyRef.

hemanth-silabs commented on include/openthread/platform/crypto.h:
```diff
@@ -185,7 +194,11 @@ typedef struct otPlatCryptoEcdsaKeyPair
+ * The public key is stored differently depending on the crypto backend library being used
+ * (OPENTHREAD_CONFIG_CRYPTO_LIB):
+ * - By default, as a byte sequence representation of an uncompressed curve point (RFC 6605 - sec 4).
+ * - When ARM PSA crypto backend is enabled, as the uncompressed representation defined by SEC1 &sect;2.3.3 as the
```
You are right here, we do need to handle this in the back end. I can update the documentation of export public key to indicate the same.

hemanth-silabs replied:
@Damian-Nordic updated the documentation to explicitly mention the export public key requirement.

Damian-Nordic replied:
Hmm, what exactly has been updated? If you decide to stick with two possible representations, doesn't it mean that SRP client must be updated to always include the RFC 6605 form?

hemanth-silabs replied:
@Damian-Nordic I have verified that both the represetations are identical. But as a precaution I have updated the documentation around `otPlatCryptoEcdsaExportPublicKey` to inform the developers that the API is expected to return the public key in RFC 6605 representation, as needed by SRP.

Damian-Nordic replied:
Are you sure? RFC 6605 states that the public key is `is a simple bit string that represents the
   uncompressed form of a curve point, "x | y"` while SEC1 in "2.3.3 Elliptic-Curve-Point-to-Octet-String Conversion" states that when the point compression is not used the EC point should be represented as `Output M = 0416 ‖ X ‖ Y `. I also noticed that Eduardo skipped the first byte returned by `psa_export_public_key` in https://github.com/zephyrproject-rtos/zephyr/blob/73ed8ccb5f2db23b89718f3cb5f0ca5142c8bb0e/modules/openthread/platform/crypto_psa.c#L485.

Damian-Nordic replied:
I see, then shouldn't this documentation here be also updated not to mention SEC1. Once exported, `otPlatCryptoEcdsaPublicKey` will only contain RFC 6605 representation. At least it was confusing to me that this doc still mentions both representations :)

hemanth-silabs replied:
I see your point, I will update this. But can you also have a look at key erasure code? I will push a update soon after.

hemanth-silabs commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -204,6 +211,105 @@ class P256
+         *
+         * @param[in] aKeyRef         PSA key reference to use while using the keypair.
+         */
+        explicit KeyPairAsRef(uint32_t aKeyRef = 0) { mKeyRef = aKeyRef; }
```
I thought I got them all :). Thanks for finding this.

hemanth-silabs commented on src/core/crypto/ecdsa.hpp:
```diff
@@ -256,6 +386,9 @@ class P256
 DefineCoreType(otPlatCryptoEcdsaKeyPair, Crypto::Ecdsa::P256::KeyPair);
 DefineCoreType(otPlatCryptoEcdsaPublicKey, Crypto::Ecdsa::P256::PublicKey);
+#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
+DefineCoreType(otPlatCryptoEcdsaKeyRef, Crypto::Ecdsa::P256::KeyPairAsRef);
```
Yes, I will remove both in my next push.

hemanth-silabs commented on include/openthread/platform/crypto.h:
```diff
@@ -650,6 +663,79 @@ otError otPlatCryptoEcdsaVerify(const otPlatCryptoEcdsaPublicKey *aPublicKey,
+ * @note This API is only used by OT core when `OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE` is enabled.
+ *
+ */
+otError otPlatCryptoEcdsaSignUsingKeyRef(otCryptoKeyRef                aKeyRef,
```
@abtink as Damian suggested, we can use `otCryptoKeyRef` instead of `otPlatCryptoEcdsaKeyRef` as ECDSA keypair is more inline with other crypto keys than network key. What are your thoughts?

abtink replied:
Agree. Sounds even better. No need to define `otPlatCryptoEcdsaKeyRef`.

hemanth-silabs commented:
@abtink @Damian-Nordic I have updated the PR to:
- Remove verifyKeyByRef inside the code.
- Updated the ECDSA to use otCryptoKey.
- Added interface to clear PSA keys when factoryresetting.

Can you please have a look?

Damian-Nordic commented on src/core/crypto/storage.hpp:
```diff
@@ -186,6 +190,12 @@ inline void DestroyKey(KeyRef aKeyRef)
 inline bool HasKey(KeyRef aKeyRef) { return otPlatCryptoHasKey(aKeyRef); }
 
+/**
+ * Delete all the persistent keys stored n PSA ITS.
```
```suggestion
 * Delete all the persistent keys stored in PSA ITS.
```

Damian-Nordic commented on src/core/mac/mac.hpp:
```diff
@@ -587,6 +587,15 @@ class Mac : public InstanceLocator, private NonCopyable
     bool IsEnabled(void) const { return mEnabled; }
 
+    /**
+     * This method indicates whether or not the link layer is enabled.
```
Copy-pasted doc.

Damian-Nordic commented on src/core/common/instance.cpp:
```diff
@@ -345,6 +345,7 @@ void Instance::Finalize(void)
     IgnoreError(otIp6SetEnabled(this, false));
     IgnoreError(otLinkSetEnabled(this, false));
 
+    Get<KeyManager>().DeleteSecurityKeys();
```
I think this call shouldn't remove persistent keys, right? Maybe make `DeleteSecurityKeys` only remove volatile keys and call `Crypto::Storage::DeletePersistentKeys` in the factory reset / erase methods only?

hemanth-silabs replied:
I wondered the same, but looking at the documentation of `otInstanceFinalize`:

Damian-Nordic replied:
Ok, maybe @abtink will help with the interpretation, but my understanding was always that `otInstanceFinalize` is used to temporarily release runtime resources taken by OT but settings and persistent data should stay as they are.

hemanth-silabs replied:
If that is the case, easily fixable :)

Damian-Nordic replied:
No, those were all my comments. PR looks good to me beside that :)

hemanth-silabs replied:
Thank you for the review, I will push the documentation updates soon.

abtink replied:
> Ok, maybe @abtink will help with the interpretation, but my understanding was always that otInstanceFinalize is used to temporarily release runtime resources taken by OT but settings and persistent data should stay as they are.

Yes, That's correct.  `otInstanceFinalize()` should not delete any saved info in non-volatile.  It just stops Thread opration and frees the current OT instance.

- We have `otInstanceErasePersistentInfo()` to erase non-voaltile settings.
- And `otInstanceFactoryReset()` which should do the same and then reset (reset platform and re-init OT instance).

hemanth-silabs replied:
Thanks for the explanation @abtink. I have updated the implementation accordingly. Please have a look.
@Damian-Nordic updated the documentation.

abtink reviewed (commented):
Thanks @hemanth-silabs. Looks good overall. 👍 
Just few smaller suggestions.

hemanth-silabs commented on src/core/thread/key_manager.cpp:
```diff
@@ -633,7 +633,23 @@ void KeyManager::SetNetworkKeyRef(otNetworkKeyRef aKeyRef)
+    mMleKey.Clear();
+    mKek.Clear();
+    Get<Mac::SubMac>().DeleteMacKeys();
+    Get<Mac::Mac>().DeleteMode2Key();
```
I am Ok either way. II chose this way, because it will be cleaner and it really doesn't affect the device behaviour if RAM keys are cleared. But if you prefer we isolate this to PSA, I am ok with it.

abtink replied:
Thanks. If we want to clear RAM, may be better to clear everything. We can do `memset(instance, 0, sizeof(otInstance))` in `Finalize` (we can consider this in a future PR).

hemanth-silabs commented on src/core/thread/key_manager.cpp:
```diff
@@ -633,7 +633,23 @@ void KeyManager::SetNetworkKeyRef(otNetworkKeyRef aKeyRef)
+    Get<Mac::SubMac>().DeleteMacKeys();
+    Get<Mac::Mac>().DeleteMode2Key();
+
+#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
```
I like this idea. I will make the changes.

hemanth-silabs commented:
@abtink @Damian-Nordic In the latest push I have addressed all your review comments. Please have a look.

abtink commented on src/core/thread/key_manager.cpp:
```diff
@@ -633,7 +633,18 @@ void KeyManager::SetNetworkKeyRef(otNetworkKeyRef aKeyRef)
 exit:
     return;
 }
+#endif // OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
+
+#if OPENTHREAD_CONFIG_PLATFORM_KEY_REFERENCES_ENABLE
+void KeyManager::DestroyTemporaryKeys(void)
```
Combine the `#if` blocks?
```suggestion

void KeyManager::DestroyTemporaryKeys(void)
```

abtink commented on src/core/thread/key_manager.hpp:
```diff
@@ -545,11 +545,25 @@ class KeyManager : public InstanceLocator, private NonCopyable
+    void DestroyTemporaryKeys(void);
+
+    /**
+     * This method deletes all the persistent keys stored in PSA ITS.
```
```suggestion
     * This method destroys all the persistent keys stored in PSA ITS.
```

abtink reviewed (approved):
LGTM. Thanks @hemanth-silabs . 👍 

Two small suggestions below:

Damian-Nordic reviewed (approved):
Looks great now!

hemanth-silabs commented:
Thanks for the reviews @abtink @Damian-Nordic!!

hemanth-silabs commented:
@jwhui Can we merge this PR?
